[4.10.0][Regression] failure when type checking with exactOptionalPropertyTypes

```
mkdir show-issue
cd show-issue
npm init -y
npm i typescript type-fest -D
npx tsc --init --skipLibCheck false --target es2022 --exactOptionalPropertyTypes
echo "import 'type-fest';" > test.ts
npx tsc --noEmit
```

See the following error:
```
node_modules/type-fest/source/merge-deep.d.ts:292:50 - error TS2344: Type '{ [KeyType in keyof ({ [Key in keyof PickIndexSignature<Options> as Key extends never ? never : Key]: PickIndexSignature<Options>[Key]; } & PickIndexSignature<...> & { [Key in keyof OmitIndexSignature<...> as Key extends "spreadTopLevelArrays" ? never : Key]: OmitIndexSignature<...>[Key]; } & OmitIndexSignature<...>...' does not satisfy the constraint '{ arrayMergeMode?: ArrayMergeMode; recurseIntoArrays?: boolean; spreadTopLevelArrays?: boolean; }'.
  Types of property 'arrayMergeMode' are incompatible.
    Type 'Options["arrayMergeMode"]' is not assignable to type 'ArrayMergeMode'.
      Type 'ArrayMergeMode | undefined' is not assignable to type 'ArrayMergeMode'.
        Type 'undefined' is not assignable to type 'ArrayMergeMode'.

292     ? MergeDeepArrayOrTuple<Destination, Source, Merge<Options, {spreadTopLevelArrays: false}>>
                                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in node_modules/type-fest/source/merge-deep.d.ts:292
```

Saw first in https://github.com/wixplosives/core3-playground/pull/175

<!-- POLAR PLEDGE BADGE START -->
## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/801">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/801/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/sindresorhus/type-fest/issues/801/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

--- a/package.json
+++ b/package.json
@@ -16,8 +16,7 @@
 		"node": ">=16"
 	},
 	"scripts": {
-		"test": "xo && tsd && tsc && npm run test:undefined-on-partial-deep && node script/test/source-files-extension.js",
-		"test:undefined-on-partial-deep": "cd test-d/undefined-on-partial-deep && tsc --project tsconfig.json"
+		"test": "xo && tsd && tsc && node script/test/source-files-extension.js"
 	},
 	"files": [
 		"index.d.ts",

--- a/source/merge-deep.d.ts
+++ b/source/merge-deep.d.ts
@@ -10,6 +10,7 @@ import type {
 	UnknownArrayOrTuple,
 } from './internal';
 import type {UnknownRecord} from './unknown-record';
+import type {EnforceOptional} from './enforce-optional';
 
 /**
 Deeply simplifies an object excluding iterables and functions. Used internally to improve the UX and accept both interfaces and type aliases as inputs.
@@ -289,7 +290,7 @@ type MergeDeepOrReturn<
 			: DefaultType
 		: Destination extends UnknownArrayOrTuple
 			? Source extends UnknownArrayOrTuple
-				? MergeDeepArrayOrTuple<Destination, Source, Merge<Options, {spreadTopLevelArrays: false}>>
+				? MergeDeepArrayOrTuple<Destination, Source, EnforceOptional<Merge<Options, {spreadTopLevelArrays: false}>>>
 				: DefaultType
 			: DefaultType>;
 

--- a/test-d/conditional-pick-deep.ts
+++ b/test-d/conditional-pick-deep.ts
@@ -12,6 +12,7 @@ interface InterfaceA {
 
 type Example = {
 	optional?: boolean;
+	optionalWithUndefined?: boolean | undefined;
 	literal: 'foo';
 	string: string;
 	map: Map<string, string>;
@@ -73,17 +74,17 @@ expectType<{
 	object: {
 		string: string;
 		subObject: {
-			optional?: string | undefined;
+			optional?: string;
 			string: string;
 		};
 	};
 }>(stringPickOptional);
 
 declare const stringPickOptionalOnly: ConditionalPickDeep<Example, string | undefined, {condition: 'equality'}>;
-expectType<{object: {subObject: {optional?: string | undefined}}}>(stringPickOptionalOnly);
+expectType<{object: {subObject: {optional?: string}}}>(stringPickOptionalOnly);
 
 declare const booleanPick: ConditionalPickDeep<Example, boolean | undefined>;
-expectType<{optional?: boolean | undefined}>(booleanPick);
+expectType<{optional?: boolean; optionalWithUndefined?: boolean | undefined}>(booleanPick);
 
 declare const numberPick: ConditionalPickDeep<Example, number>;
 expectType<{number: 1; interface: {a: number}}>(numberPick);

--- a/test-d/undefined-on-partial-deep.ts
+++ b/test-d/undefined-on-partial-deep.ts
@@ -1,8 +1,6 @@
-/**
-@note This file is used for testing by `tsc` but not `tsd`, so we can just test assignable.
-*/
+// TODO: Test equality
 import {expectAssignable} from 'tsd';
-import type {UndefinedOnPartialDeep} from '../../source/undefined-on-partial-deep';
+import type {UndefinedOnPartialDeep} from '../source/undefined-on-partial-deep';
 
 type TestType1 = UndefinedOnPartialDeep<{required: string; optional?: string; optional2?: number; optional3?: string}>;
 expectAssignable<TestType1>({required: '', optional: undefined, optional2: 1});

--- a/test-d/undefined-on-partial-deep/tsconfig.json
+++ b/test-d/undefined-on-partial-deep/tsconfig.json
@@ -1,7 +1,0 @@
-{
-	"extends": "../../tsconfig",
-	"compilerOptions": {
-		"exactOptionalPropertyTypes": true,
-	},
-	"exclude": []
-}

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -7,6 +7,7 @@
 			"ES2020",
 			"DOM"
 		],
+		"exactOptionalPropertyTypes": true,
 		"skipLibCheck": false, // Ensures .d.ts files are checked: https://github.com/sindresorhus/tsconfig/issues/15
 	},
 	"exclude": [

